src/Storages/MaterializedView/RefreshTask.cpp

```diff
@@ -1008,41 +1009,92 @@ void RefreshTask::interruptExecution()
+    ///     If this fails, retry until we see a different table by the same name.
+
+    StoragePtr prev_storage;
+    /// (Without maybe_unused, clang-tidy-19 seems to produce incorrect error
+    ///  "Value stored to 'prev_table_dropped_locally' is never read" about the assignment
+    ///  prev_table_dropped_locally = false further down.)
+    [[maybe_unused]] bool prev_table_dropped_locally = false;
+    std::exception_ptr exception;
+
     for (int attempt = 0; attempt < 10; ++attempt)
     {
-        StoragePtr prev_storage = std::move(storage);
-        storage = DatabaseCatalog::instance().getTable(storage_id, context);
+        if (attempt > 0 && !prev_table_dropped_locally)
+        {
+            /// We're waiting for DatabaseReplicated to catch up and see the new table.
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        }
+
+        StoragePtr storage = DatabaseCatalog::instance().getTable(storage_id, context);
+
         if (storage == prev_storage)
         {
-            // Table was dropped but is still accessible in DatabaseCatalog.
-            // Either ABA problem or something's broken. Don't retry, just in case.
-            break;
+            if (prev_table_dropped_locally)
+                // Table was dropped but is still accessible in DatabaseCatalog.
+                // Either ABA problem or something's broken. Don't retry.
+                break;
+            continue;
         }
-        storage_lock = storage->tryLockForShare(context->getCurrentQueryId(), context->getSettingsRef()[Setting::lock_acquire_timeout]);
-        if (storage_lock)
-            break;
-    }
-    if (!storage_lock)
-        throw Exception(ErrorCodes::TABLE_IS_DROPPED, "Table {} is dropped or detached", storage_id.getFullNameNotQuoted());
+        prev_storage = storage;
 
-    if (coordination.coordinated)
-    {
-        UUID uuid = storage->getStorageID().uuid;
+        TableLockHolder storage_lock = storage->tryLockForShare(context->getCurrentQueryId(), context->getSettingsRef()[Setting::lock_acquire_timeout]);
+        if (!storage_lock)
+        {
+            prev_table_dropped_locally = true;
+            continue;
+        }
+        prev_table_dropped_locally = false;
 
-        std::lock_guard lock(replica_sync_mutex);
-        if (uuid != last_synced_inner_uuid)
+        if (coordination.coordinated)
         {
-            InterpreterSystemQuery::trySyncReplica(storage, SyncReplicaMode::DEFAULT, {}, context);
+            std::lock_guard lock(replica_sync_mutex);
+            UUID uuid = storage->getStorageID().uuid;
+            if (uuid != last_synced_inner_uuid)
+            {
+                try
+                {
+                    InterpreterSystemQuery::trySyncReplica(storage, SyncReplicaMode::DEFAULT, {}, context);
+                }
+                catch (Exception & e)
+                {
+                    if (e.code() != ErrorCodes::ABORTED)
```

Comment:
Shouldn't we retry only specific error? Or at least do not retry `LOGICAL_ERROR`

Reply:
I think I'll keep ABORTED, don't want to hunt down all code paths that throw it in RMT and SMT.